### PR TITLE
Refactoring the action API handler

### PR DESCRIPTION
**Pull request by rabbah**
This pr refactors the action API handler and the entitlement provider:
- the action activation bits are factored into their own traits and enable mock testing of the "invoke action" functionality. It also makes invoking an action accessible from other API handlers that need to post activations.
- the entitlement checks to permit checking of quotas independent of resources.

### CODE REVIEW

**Comment by rabbah** [comment]
pg2/737 has landed.

File: common/scala/src/main/scala/whisk/core/entity/FullyQualifiedEntityName.scala [hunk 1]
```diff
@@ -31,11 +31,25 @@ import whisk.core.entity.size.SizeString
  */
 protected[core] case class FullyQualifiedEntityName(path: EntityPath, name: EntityName, version: Option[SemVer] = None) extends ByteSizeable {
     private val qualifiedName: String = path + EntityPath.PATHSEP + name
+    /** Resolves default namespace in path to given name if the root path is the default namespace. */
     def resolve(namespace: EntityName) = FullyQualifiedEntityName(path.resolveNamespace(namespace), name, version)
+
+    /** @return full path including name, i.e., "path/name" */
+    def fullpath: EntityPath = path.addpath(name)
```
Comment (psuter) [inline common/scala/src/main/scala/whisk/core/entity/FullyQualifiedEntityName.scala]: Any reason `fullpath` and `addpath` aren't camel-cased?

File: core/controller/src/main/scala/whisk/core/entitlement/Entitlement.scala [hunk 2]
```diff
@@ -225,12 +247,17 @@ protected[core] abstract class EntitlementProvider(config: WhiskConfig, loadBala
...
 
-    /** Limits activations if the load balancer is overloaded. */
-    protected def checkSystemOverload(subject: Subject, right: Privilege)(implicit transid: TransactionId) = {
+    /**
+     * Limits activations if the system is overloaded.
+     *
+     * @param right the privilege, if ACTIVATE then check quota else return None
+     * @return None if system is not overloaded else a rejection
+     */
+    protected def checkSystemOverload(right: Privilege)(implicit transid: TransactionId) = {
```
Comment (psuter) [inline core/controller/src/main/scala/whisk/core/entitlement/Entitlement.scala]: Writing down the return type helps the reviewers :)

File: core/controller/src/main/scala/whisk/core/entitlement/Entitlement.scala [hunk 3]
```diff
@@ -239,16 +266,26 @@ protected[core] abstract class EntitlementProvider(config: WhiskConfig, loadBala
...
             val isTrigger = resources.exists(_.collection.path == Collection.TRIGGERS)
             (isInvocation && !invokeRateThrottler.check(subject)) || (isTrigger && !triggerRateThrottler.check(subject))
         }
+        hasTooManyRequests(right == ACTIVATE && userThrottled)
+    }
 
-        if (right == ACTIVATE && userThrottled) {
-            Some(Future.failed(RejectRequest(TooManyRequests, tooManyRequests)))
+    /** Helper. */
+    private def hasTooManyRequests(yes: Boolean)(implicit transid: TransactionId) = {
```
Comment (psuter) [inline core/controller/src/main/scala/whisk/core/entitlement/Entitlement.scala]: Name of function is a little misleading. Maybe:
```
private def requestCheckToRejection(hasTooMany: Boolean)(...) : Option[..] = { ... }
```
?

File: core/controller/src/main/scala/whisk/core/entitlement/Entitlement.scala [hunk 4]
```diff
@@ -257,15 +294,25 @@ protected[core] abstract class EntitlementProvider(config: WhiskConfig, loadBala
...
             val isInvocation = resources.exists(_.collection.path == Collection.ACTIONS)
             (isInvocation && !concurrentInvokeThrottler.check(subject))
         }
+        hasTooManyConcurrentRequests(right == ACTIVATE && userThrottled)
+    }
 
-        if (right == ACTIVATE && userThrottled) {
-            Some(Future.failed(RejectRequest(TooManyRequests, tooManyConcurrentRequests)))
+    /** Helper. */
+    private def hasTooManyConcurrentRequests(yes: Boolean)(implicit transid: TransactionId) = {
```
Comment (psuter) [inline core/controller/src/main/scala/whisk/core/entitlement/Entitlement.scala]: same as above (and can probably be refactored into a single one)

File: core/controller/src/main/scala/whisk/core/entitlement/Entitlement.scala (hunk 2 above)
Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/entitlement/Entitlement.scala]: The happiness of my reviewer(s) is very important to me. https://github.com/openwhisk/openwhisk/pull/1613/commits/9865aa5aff49b7ebf5ad8873f433f759a549a4f8

File: common/scala/src/main/scala/whisk/core/entity/FullyQualifiedEntityName.scala (hunk 1 above)
Comment (rabbah) [inline common/scala/src/main/scala/whisk/core/entity/FullyQualifiedEntityName.scala]: Turned into a bit more: https://github.com/openwhisk/openwhisk/pull/1613/commits/998b43e2d87d35a237ad8bdf957d0573c9680003

**Comment by rabbah** [comment]
pg2/745 approved.
going to rebase and squash.

**Comment by rabbah** [comment]
Going to close since #1600 is a superset.

**Pull request closed without merging**